### PR TITLE
Reduce peak memory allocation in eager loop Adam, AdamW, NAdam, and Rprop

### PR DESCRIPTION
**Pull request by Vic-41148**
Made for #104789

After checking _single_tensor_* for-loop optimizers there were spots in capturable, differentiable branches where out-of-place ops were allocating temporaries that could instead write in-place, all changes gated on differentiable=False cuz autograd needs the out-of-place graph for backwards

For each of them:

 - Adam: torch.maximum(..., out=) for the amsgrad max update instead of clone+copy, denom computed as a chained .sqrt().div_().add_() in-place instead of using inbetween tensor

 - NAdam: just changed .sqrt_() instead of .sqrt() on the denom.

 - Rprop: used .sign_() instead of .sign(), collapse three sequential torch.where/copy_ calls into one nested torch.where,used masked_fill_ instead of copy_(torch.where(...)) for grad zero

Adam capturable path (no amsgrad, no weight_decay) drops one param sized intermediate, before patch mem peak was 11264 bytes, and now after patch is 7680 bytes (CUDA, N=1024 float32), with our tight budget of 8192 bytes the old code was failing that budget, the new code passes it fine

NAdam denom chain switch .sqrt() to .sqrt_(), removed one temporary there, but this does not move measured peak memory (14848 bytes before and after patch)

NAdam step peak is somewhere else in the sequence temporary does not change it, no mem regression test is tied to NAdam peak number because of that

This change is to be consistent with the other capturable-path work done, not saving any mem here

Rprop changes work on already allocated tensors, so that's verified by its output compared to before the patch, not by peak memory

Added tests for this in test/test_optim.py using  @optims(optim_db)

Eager blind spot:
 - The original broken rprop.py (with unconditional .sign_() outside the differentiable branch) passed test_diff_gradcheck_rprop in eager mode in every test I did  and both fast_mode settings, eager gradcheck gave zero regression protection for this type of bug, but the dynamo CI shard did catch it
 
 Masked_fill_ gating:
  - The condition capturable and not differentiable for grad.masked_fill_() in Rprop is to not use in-place ops on the differentiable path, no existing optim_db config is testing capturable=True, differentiable=True together for Rprop here, so this is currently untested by the parametrized suite
  
 Dynamo skips (thse were pre existing not caused by this PR):
 
  Three gradcheck tests are being skipped in PYTORCH_TEST_WITH_DYNAMO=1 due to pre existing compiled_autograd on main:
  
  - Adam,NAdam: gradchecks check_undefined_grad_support triggers compiled_autograds validate_outputs to return None for a undefined-grad probe, which is flowing in torch.add(tensor, None) and then crashes with a typeError, confirmed it is reproducible on main with all my optimizer files reverted back to as it was before
  
  - Rprop: dynamo when compiled backward produced a incorrect gradient for the differentiable path under gradcheck, this one also confirmed reproducible on main, this was not caused by this PR it only did in-place rewrites

### CODE REVIEW

File: torch/optim/nadam.py [hunk 1]
```diff
@@ -353,7 +353,7 @@ def _single_tensor_nadam(
         # decay the first and second moment running average coefficient
         exp_avg.lerp_(grad, 1 - beta1)
         exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
-        denom = exp_avg_sq.div(bias_correction2).sqrt()
+        denom = exp_avg_sq.div(bias_correction2).sqrt_()
```
Comment (Skylion007) [inline torch/optim/nadam.py]: Reason we don't do rsqrt here? ULP error I am guessing?
